service/share/share.go

```diff
@@ -0,0 +1,124 @@
+package share
+
+import (
+	"context"
+
+	format "github.com/ipfs/go-ipld-format"
+	"github.com/ipfs/go-merkledag"
+
+	"github.com/celestiaorg/nmt/namespace"
+
+	"github.com/celestiaorg/celestia-node/ipld"
+	"github.com/celestiaorg/celestia-node/ipld/plugin"
+	"github.com/celestiaorg/celestia-node/service/header"
+)
+
+// TODO(@Wondertan): We prefix real data of shares with namespaces to be able to recover them during erasure coding
+//  recovery. However, that is storage and bandwidth overhead(8 bytes per each share) which we can avoid by getting
+//  namespaces from CIDs stored in IPLD NMT Nodes, instead of encoding namespaces in erasure coding.
+//
+// TODO(@Wondertan): Ideally, Shares structure should be defined in a separate repository with
+//  rsmt2d/nmt/celestia-core/celestia-node importing it. rsmt2d and nmt are already dependent on the notion of "share",
+//  so why shouldn't we have a separated and global type for it to avoid the type mess with defining own share type in
+//  each package.
+//
+// Share is
+type Share = namespace.PrefixedData8
+
+// Service provides as simple interface to access any DataSquare/Block Share on the network.
+//
+// All Get methods follow the following flow:
+// 	* Check local storage for the requested Share.
+// 		* If exist
+// 			* Load from disk
+//			* Return
+//  	* If not
+//  		* Find provider on the network
+//      	* Fetch the Share from the provider
+//			* Store the Share
+//			* Return
+type Service interface {
+	// GetShare loads a Share committed to the given DataAvailabilityHeader by its Row anc Col coordinates in the
+	// erasures data square or block.
+	GetShare(ctx context.Context, dah header.DataAvailabilityHeader, row, col int) (Share, error)
```

Comment:
should we pass pointer to DAH for these `GET` methods?

Reply:
Similarly, could context be passed as a reference?

Reply:
> Similarly, could context be passed as a reference?

@adlerjohn, Context is an interface, so it is passed as a pointer by default. That's a Golang axiom.

Reply:
@renaynay, yeah, having DAH as pointer seems more or less reasonable

Reply:
Fixed in 99f1f11